### DISCUSSION TITLE
Testing Plan for Stacks 2.4

### DISCUSSION THREAD

**Discussion by saralab** [stacks-network/stacks-core#3702] [General]
This document will be edited as more test items are accepted.

## Blockchain track
- [x] Verify that the `.pox-3` activation works on testnet - @deantchi @kantai 
  - [x] `.pox-3` added to `contract_versions` in `/v2/pox` RPC response
  - [x] `.pox-3` instantiates at the expected block height [Regtest integration tests](https://github.com/stacks-network/stacks-blockchain/pull/3715)
  - [x] Users can stack their STX into `.pox-3` [Regtest integration tests](https://github.com/stacks-network/stacks-blockchain/pull/3715)
  - [x] Once the next reward cycle begins, tokens stacked in `.pox-3` are used to derive the reward cycle. `.pox-2` is ignored [Regtest integration tests](https://github.com/stacks-network/stacks-blockchain/pull/3715)
  - [x] 2.4 Epoch transitions successfully on Testnet - Block #2432545 
- [x] Verify correct `.pox-3` behavior on devnet & testnet- @obycode - In Progress - fully tested on devnet; testnet testing pending
  - [x] `stack-increase` works as expected
  - [x] `stack-extend` works as expected
  - [x] `delegate-stx` works as expected
  - [x] `revoke-delegate-stx` works as expected
  - [x] `delegate-stack-stx` works as expected
  - [x] `delegate-stack-increase` works as expected
  - [x] `delegate-stack-extend` works as expected
  - [x] `stack-aggregation-commit-indexed` works as expected
  - [x] verify that btc wire format tx for "delegate-stx" works
  - [x]  `stack-unlock` (auto-unlocks) works as expected

- [x] Verify that all tests in https://github.com/hirosystems/stacks-2-1-testing still work correctly - @obycode
- [ ] 100% Clarity code coverage with clarinet unit tests - @jbencin - In Progress
  - [x] Fork https://github.com/Trust-Machines/pox-clarinet-tests and modify for PoX3
  - [x] Make a replacement for `stx-account` that works under Clarinet (works except when delegating)
  - [x] Fix existing tests
  - [x] Reach as close as possible to 100% coverage (currently at 73%)
    - This is likely as far as we can get under Clarinet. Getting `delegate-` would require modifying the contract to a point that would make any testing on it not very meaningful

- [x] Stacks API works correctly with Epoch 2.4 @zone117x 
  - [x] Pox-3 events get emitted, should be same behavior as pox-2 events
  - [x] Rosetta stacking operations uses pox-3
- [x] Fuzz testing - @fpbgg - In Progress
  - [x] No inconsistencies between Clarity implementation
  - [x] Fuzz testing ongoing for Clarity
## Dev tooling track - @janniks
- [x] Stacks.js `stacking` library uses pox-3 automatically

# Non-blocking items
## Subnets
- [ ] Verify that subnets instantiate on the public 2.4 testnet - @jbencin  - In Progress
  - [x] Tested and failed. Issue has been tracked to [this](https://github.com/hirosystems/stacks-subnets/commit/50027920222f8acbdf37c155cfbc5ffc0f4efd79) commit, which changes the interface of `commit-block` in `subnet.clar`
  - [ ] Update subnet contracts on testnet L1 and L2 so that they match and test again
 
## Pools 
- [ ] Ensure pools work as expected with `.pox-3`

## Client/Wallet track
- [ ] Clients that stack will automatically direct STX to `.pox-3`
  - [ ] Clients can determine that STX have unlocked after the transition
  - [ ] Clients can determine when the `.pox-3` contract has been instantiated
- [ ] Clients will report the correct locked / unlocked balance after the transition
  - [ ] At the transition, all STX locked in `.pox-2` show up as unlocked
  - [ ] STX stacked after the transition show up as locked
## Exchange track - @wileyj | @cuevasm
- [ ] Reach out to all exchanges we know of to inform them that they will need to update their code to stack user STX to `.pox-3`
- [ ] Offer them a testnet (or a private testnet) to try out new stacking
- [ ] Inform exchanges that they should use an officially-supported client library for `.pox-3` if the are able
---
Please suggest more tests below

**Comment by cargocultscience** [comment]
Very happy to see this kind of effort and documentation!

**Comment by saralab** [comment]
Update May 16 2023: 
- Private Testnet ready to test the stacking scenarios for PoX-3, stacks-increase and stacks-extend look good so far

**Comment by whoabuddy** [comment]
Tagging @yknl @dantrevino @aulneau for tooling/client/wallet topics, my guess is they are already clued in but just want to be sure they've seen this and can share with others:slightly_smiling_face: 

> verify that btc wire format tx for "delegate-stx" works

should this include a verification of the btc wire format for "stack-stx" too?
